crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -568,6 +568,8 @@ where
 
                     while let Some((index, (piece_index, result))) = pieces_stream.next().await {
                         debug!(%batch, %index, %piece_index, "Downloaded piece");
+                        // Release slot for future batches (dropping the split permit releases the slot).
+                        permit.split(1);
```

Comment:
I actually think it'd be better to store this in a variable and let it drop naturally once it goes out of scope:
```suggestion
                        let _permit = permit.split(1);
```

Otherwise this opens a possibility to download more pieces than expected and running out of memory. This was also an issue before, but I think it makes sense to handle it now that the code has moved.

Reply:
Done!